Added documentation for dataset mapping customization. (#221)
ubccr/xdmod-supremm: docs/extending.md

---
title: Advanced Usage
---

This document describes how to modify the configuration files to show data from
different datasources than the default. 

> However reuqires detailed knowledge. Also if you modify /extend and

Reviewer: ```suggestion
However, this requires detailed knowledge. Also if you modify/extend
```

```
/usr/share/xdmod/etl/js/config/supremm/dataset_maps/pcp.js
```

XDMoD supports using a different dataset mapping file for each
HPC resource. THe dataset mapping file that is used is configured
in the `datasetmap` parameter in the [supremm_resources.json](supremm-configuration.html#supremm_resourcesjson)
file.  The dataset mapping is a written in javascript and supports
arbitrary javascript functions to transform the job data.

## Example

The following example shows how to modify the dataset mapping file to support
reading job data from the job summary document that is different from the
default. 

In this example we describe how to extend the dataset mapping to map data
from the lustre llite statistics to the Filesystem I/O statistics in XDMoD.
The default mapping file reads nfs filesystem metrics based on the configuration
settings in the [supremm_resources.json](supremm-configuration.html#supremm_resourcesjson) file. This example
shows how to override this so the metrics are instead gathered from the lustre metrics.

The steps to configure the software are as follows:
1. Create a new dataset map.
1. Configure XDMoD to use the new dataset map.

> 1. Test the new data map.

Reviewer: ```suggestion
1. Test the new dataset map.
```

1. Re-ingest the job data into XDMoD.

### 1. Create a new dataset mapping

A new dataset mapping file should be created in `/usr/share/xdmod/etl/js/config/supremm/dataset_maps`. In this

> example, the new mapping file will be called `custom.js`. The file name should follow the nodejs

Reviewer: ```suggestion
example, the new mapping file will be called `custom.js`. The file name should follow the Node.js
```

module naming convention. The source of the new mapping file is shown below:
```js
// load the default mapping module
var pcp = require('./pcp.js');

// load the library that contains helper functions
var map_helpers = require('../map_helpers.js');

module.exports = function (config) {

    // create a new copy of the default map
    var pcp_map = new pcp(config);

    // override the mapping attributes for netdir home and scratch:
    
    pcp_map.attributes.netdir_home_read = map_helpers.device('lustre', '/home', 'read_bytes');
    pcp_map.attributes.netdir_home_write = map_helpers.device('lustre', '/home', 'write_bytes');

    pcp_map.attributes.netdir_scratch_read = map_helpers.device('lustre', '/scratch', 'read_bytes');
    pcp_map.attributes.netdir_scratch_write = map_helpers.device('lustre', '/scratch', 'write_bytes');

    // can add more overrides here....

    return pcp_map;
}
```
This loads the default mapping and selectively overrides fields.

### 2. Configure XDMoD to use the new map.

> The supremm_resources.json](supremm-configuration.html#supremm_resourcesjson) file defines

Reviewer: ```suggestion
The [supremm_resources.json](supremm-configuration.html#supremm_resourcesjson) file defines
```
